constraint/pkg/client/rego_helpers.go

```diff
@@ -0,0 +1,95 @@
+package client
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/open-policy-agent/opa/ast"
+)
+
+// ensureRegoConformance rewrites the package path and ensures there is no access of `data`
+// beyond the whitelisted bits. Note that this rewriting will currently modify the Rego to look
+// potentially very different from the input, but it will still be functionally equivalent.
+func ensureRegoConformance(kind, path, rego string) (string, error) {
+	if rego == "" {
+		return "", errors.New("Rego source code is empty")
+	}
+	module, err := ast.ParseModule(kind, rego)
+	if err != nil {
+		return "", err
+	}
+	if err := rewritePackage(path, module); err != nil {
+		return "", err
+	}
+	if len(module.Imports) != 0 {
+		return "", errors.New("Use of the `import` keyword is not allowed")
+	}
+	if err := checkDataAccess(module); err != nil {
+		return "", err
+	}
+	return module.String(), nil
+}
+
+// rewritePackage rewrites the `package` statement in Rego with the provided package path
+func rewritePackage(path string, module *ast.Module) error {
+	if module.Package == nil {
+		return errors.New("No `package` statement parsed")
+	}
+	pathParts := strings.Split(path, ".")
+	packageRef := ast.EmptyRef()
+	if pathParts[0] != "data" {
+		packageRef = append(packageRef, ast.NewTerm(ast.String("data")))
+	}
+	for _, v := range pathParts {
+		packageRef = append(packageRef, ast.NewTerm(ast.String(v)))
+	}
+	module.Package.Path = packageRef
+	return nil
+}
+
+func makeInvalidRootFieldErr(val ast.Value, allowed map[string]bool) string {
+	var validFields []string
+	for field := range allowed {
+		validFields = append(validFields, field)
+	}
+	return fmt.Sprintf("Invalid `data` field: %s. Valid fields are: %s", val.String(), strings.Join(validFields, ", "))
+}
+
+// checkDataAccess makes sure that data is only referenced in terms of valid subfields
+func checkDataAccess(module *ast.Module) error {
+	// Currently rules should only access data.inventory
+	validDataFields := map[string]bool{
+		"inventory": true,
+	}
+
+	var errs []string
```

Comment:
Nit: alternatively we could define something like `type Errors []error` that implements `Error() string`. This has the benefit of preserving the error values for the caller.

Reply:
done